endpoints/put.go

```diff
@@ -12,133 +13,252 @@ import (
 	"github.com/julienschmidt/httprouter"
 	"github.com/prebid/prebid-cache/backends"
 	backendDecorators "github.com/prebid/prebid-cache/backends/decorators"
+	"github.com/prebid/prebid-cache/metrics"
 	"github.com/prebid/prebid-cache/utils"
 	"github.com/sirupsen/logrus"
 )
 
 // PutHandler serves "POST /cache" requests.
-func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) func(http.ResponseWriter, *http.Request, httprouter.Params) {
-	// TODO(future PR): Break this giant function apart
-	putAnyRequestPool := sync.Pool{
-		New: func() interface{} {
-			return &PutRequest{}
-		},
+type PutHandler struct {
+	backend backends.Backend
+	cfg     putHandlerConfig
+	memory  syncPools
+	metrics *metrics.Metrics
+}
+
+type putHandlerConfig struct {
+	maxNumValues int
+	allowKeys    bool
+}
+
+type syncPools struct {
+	requestPool     sync.Pool
+	putResponsePool sync.Pool
+}
+
+func NewPutHandler(storage backends.Backend, metrics *metrics.Metrics, maxNumValues int, allowKeys bool) func(http.ResponseWriter, *http.Request, httprouter.Params) {
+	putHandler := &PutHandler{}
+
+	// Assign storage client to put endpoint
+	putHandler.backend = storage
+
+	// pass metrics engine
+	putHandler.metrics = metrics
+
+	// Pass configuration values
+	putHandler.cfg = putHandlerConfig{
+		maxNumValues: maxNumValues,
+		allowKeys:    allowKeys,
 	}
 
-	putResponsePool := sync.Pool{
-		New: func() interface{} {
-			return &PutResponse{}
+	// Instantiate thread-safe memory pools
+	putHandler.memory = syncPools{
+		requestPool: sync.Pool{
+			New: func() interface{} {
+				return &PutRequest{}
+			},
+		},
+		putResponsePool: sync.Pool{
+			New: func() interface{} {
+				return &PutResponse{}
+			},
 		},
 	}
 
-	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-		body, err := ioutil.ReadAll(r.Body)
-		if err != nil {
-			http.Error(w, "Failed to read the request body.", http.StatusBadRequest)
-			return
-		}
-		defer r.Body.Close()
+	return putHandler.handle
+}
 
-		put := putAnyRequestPool.Get().(*PutRequest)
-		put.Puts = make([]PutObject, 0)
-		defer putAnyRequestPool.Put(put)
+// parseRequest unmarshals the incoming put request into a thread-safe memory pool
+// If the incoming request could not be unmarshalled or if the request comes with more
+// elements to put than the maximum allowed in Prebid Cache's configuration, the
+// corresponding error is returned
+func (e *PutHandler) parseRequest(r *http.Request) (*PutRequest, error) {
+	if r == nil {
+		return nil, utils.PutBadRequestError{}
+	}
 
-		err = json.Unmarshal(body, put)
-		if err != nil {
-			http.Error(w, "Request body "+string(body)+" is not valid JSON.", http.StatusBadRequest)
-			return
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, utils.PutBadRequestError{}
```

Comment:
:thinking:   Not sure. If there was an unmarshal error, we are adding it but the current version of PBC in production is not doing it from here. From the library's comment below:
```
 72 // Callers should always process the n > 0 bytes returned before
 73 // considering the error err. Doing so correctly handles I/O errors
 74 // that happen after reading some bytes and also both of the
 75 // allowed EOF behaviors.
 76 //
 77 // Implementations of Read are discouraged from returning a
 78 // zero byte count with a nil error, except when len(p) == 0.
 79 // Callers should treat a return of 0 and nil as indicating that
 80 // nothing happened; in particular it does not indicate EOF.
 81 //
 82 // Implementations must not retain p.
 83 type Reader interface {
 84     Read(p []byte) (n int, err error)
 85 }
/usr/local/go/src/io/io.go
```
...seems that it would be ok to do so. Because they encourage to consider the n > 0 bytes before the error. We probably need to look further into this before considering this check, and maybe we can add those hypothetical changes to a future PR if that's ok with you.